import yaml

headers_dict = [
    "SAMPLE",
    "HGVS",
    "GENE/LOCUS",
    "GENE/LOCUS DESCRIPTION",
    "TOTAL LOCUS DEPTH",
    "VARIANT HETEROPLASMY",
    "ALT DEPTH",
    "REF DEPTH",
    "TOTAL SAMPLE DEPTH",
    "VARIANT QUALITY",
    "TIER",
    "COHORT COUNT",
    "COHORT FREQUENCY",
    "CHR",
    "POS",
    "REF",
    "ALT",
    "QUAL",
    "FILTER",
    "MQM_INFO",
    "MQMR_INFO",
    "QA_INFO",
    "QR_INFO",
    "SAF_INFO",
    "SAR_INFO",
    "SRF_INFO",
    "SRR_INFO",
    "SBR_INFO",
    "SBA_INFO",
    "POS_FILTER",
    "SBR_FILTER",
    "SBA_FILTER",
    "MQMR_FILTER",
    "AQR_FILTER",
    "COMMERCIAL PANELS",
    "PHYLOTREE HAPLOTYPE",
    "MITOTIP SCORE",
    "MITOTIP PERCENTILE",
    "MITOTIP INTERPRETATION",
    "ANTICODON",
    "ALLELE FREQUENCY MITOMAP",
    "DISEASE MITOMAP",
    "MGRB FREQUENCY",
    "MGRB FILTER",
    "MGRB AC",
    "MGRB AN",
    "PHYLOTREE MUT",
    "LOCUS MITOMAP",
    "NUMBER OF REFERENCES MITOMAP",
    "VARIANT AMINO ACID CHANGE MITOMAP",
    "CODON POSITION MITOMAP",
    "CODON NUMBER MITOMAP",
    "NUM DISEASE REFERENCES MITOMAP",
    "RNA MITOMAP",
    "HOMOPLASMY MITOMAP",
    "HETEROPLASMY MITOMAP",
    "STATUS MITOMAP",
    "DISEASE AMINO ACID CHANGE MITOMAP",
    "GENBANK FREQUENCY MITOMAP",
    "GT_FORMAT",
    "QR_FORMAT",
    "AQR_FORMAT",
    "QA_FORMAT",
    "AQA_FORMAT",
    "INFO",
    "FORMAT",
]

vep_excel_headers = [
    "HIGHEST VEP IMPACT",
    "CONSEQUENCE VEP",
    "CODONS VEP",
    "AMINO ACIDS VEP",
    "GENE VEP",
    "SYMBOL VEP",
    "FEATURE VEP",
    "EXON VEP",
    "POLYPHEN VEP",
    "SIFT VEP",
    "PROTEIN POSITION VEP",
    "BIOTYPE VEP",
    "CANONICAL VEP",
    "FEATURE TYPE VEP",
    "cDNA POSITION VEP",
    "CDS POSITION VEP",
    "EXISTING VARIATION VEP",
    "DISTANCE VEP",
    "STRAND VEP",
    "CLIN SIG VEP",
    "LOF FLAGS VEP",
    "LOF FILTER VEP",
    "LOF VEP",
    "RADIALSVM SCORE VEP",
    "RADIALSVM PRED VEP",
    "LR SCORE VEP",
    "LR PRED VEP",
    "CADD RAW VEP",
    "CADD PHRED VEP",
    "RELIABILITY INDEX VEP",
    "HGVSC VEP",
    "HGVSP VEP",
]


vcf_column_names = {
    "info": {
        "Map_Locus": "GENE/LOCUS",
        "Description": "GENE/LOCUS DESCRIPTION",
        "DP": "TOTAL LOCUS DEPTH",
        "MQM": "MQM_INFO",
        "MQMR": "MQMR_INFO",
        "QA": "QA_INFO",
        "QR": "QR_INFO",
        "SAF": "SAF_INFO",
        "SAR": "SAR_INFO",
        "SRF": "SRF_INFO",
        "SRR": "SRR_INFO",
        "SBR": "SBR_INFO",
        "SBA": "SBA_INFO",
    },
    "format": {
        "VAF": "VARIANT HETEROPLASMY",
        "AO": "ALT DEPTH",
        "RO": "REF DEPTH",
        "DP": "TOTAL SAMPLE DEPTH",
        "q": "VARIANT QUALITY",
        "tier": "TIER",
        "GT": "GT_FORMAT",
        "QR": "QR_FORMAT",
        "AQR": "AQR_FORMAT",
        "QA": "QA_FORMAT",
        "AQA": "AQA_FORMAT",
        "POS_filter": "POS_FILTER",
        "SBR_filter": "SBR_FILTER",
        "SBA_filter": "SBA_FILTER",
        "MQMR_filter": "MQMR_FILTER",
        "AQR_filter": "AQR_FILTER",
    },
    "annotations": {
        "commercial_panels": "COMMERCIAL PANELS",
        "phylotree_haplotype": "PHYLOTREE HAPLOTYPE",
        "MitoTip_score": "MITOTIP SCORE",
        "MitoTip_percentile": "MITOTIP PERCENTILE",
        "MitoTip_interpretation": "MITOTIP INTERPRETATION",
        "anticodon": "ANTICODON",
        "disease_mitomap": "DISEASE MITOMAP",
        "MGRB_frequency": "MGRB FREQUENCY",
        "MGRB_FILTER": "MGRB FILTER",
        "MGRB_AC": "MGRB AC",
        "MGRB_AN": "MGRB AN",
        "phylotree_mut": "PHYLOTREE MUT",
        "locus_mitomap": "LOCUS MITOMAP",
        "num_references_mitomap": "NUMBER OF REFERENCES MITOMAP",
        "variant_amino_acid_change_mitomap": "VARIANT AMINO ACID CHANGE MITOMAP",
        "codon_position_mitomap": "CODON POSITION MITOMAP",
        "codon_number_mitomap": "CODON NUMBER MITOMAP",
        "num_disease_references_mitomap": "NUM DISEASE REFERENCES MITOMAP",
        "RNA_mitomap": "RNA MITOMAP",
        "homoplasmy_mitomap": "HOMOPLASMY MITOMAP",
        "heteroplasmy_mitomap": "HETEROPLASMY MITOMAP",
        "status_mitomap": "STATUS MITOMAP",
        "disease_amino_acid_change_mitomap": "DISEASE AMINO ACID CHANGE MITOMAP",
        "GenBank_frequency_mitomap": "GENBANK FREQUENCY MITOMAP",
    },
}

excel_headers = headers_dict

data = {
    "excel_headers": excel_headers,
    "vep_excel_headers": vep_excel_headers,
    "vcf_headers": vcf_column_names,
}

if __name__ == "__main__":
    with open("report-config.yaml", "w") as file:
        yaml.dump(data, file)
