# Mity report documentation

| **Term**                      | **Definition/Info**                                                                                                                                                                                                                                                                                                                                                                                     |
| ----------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **HGVS**                      | Variant annotation using HGVS syntax, which assumes variant is SNP, insertion, or deletion.                                                                                                                                                                                                                                                                                                             |
| **gene/locus**                | The gene that the variant falls in, or if it is not in a gene, the mitomap entry, if it exists.                                                                                                                                                                                                                                                                                                         |
| **variant_heteroplasmy**      | alt_depth/(ref_depth+alt_depth)                                                                                                                                                                                                                                                                                                                                                                         |
| **variant_quality**           | Phred-scaled variant quality                                                                                                                                                                                                                                                                                                                                                                            |
| **total_sample_depth**        | alt_depth + ref_depth for the particular sample                                                                                                                                                                                                                                                                                                                                                         |
| **total_locus_depth**         | Sum of all the total_sample_depth in the VCF file                                                                                                                                                                                                                                                                                                                                                       |
| **COHORT_COUNT**              | The number of times that variant (defined by position and alternate) is seen in all of the input VCFs.                                                                                                                                                                                                                                                                                                  |
| **tier**                      | 1:=VAF>=1%; 2:=VAF<1% and alt_depth>=10; 3:=VAF<1% and alt_depth<10                                                                                                                                                                                                                                                                                                                                     |
| **commercial_panels**         | Indicates whether a variant is commonly tested in commercial/academic testing laboratories. Baylor: [link](https://www.bcm.edu/research/medical-genetics-labs/test_detail.cfm?testcode=2010); Common22: [link](https://www.vcgs.org.au/order/tests/634); Common58: [link](https://www.genedx.com/test-catalog/available-tests/58-confirmed-disease-causing-mtdna-point-mutations-and-deletion-testing/) |
| **MGRB**                      | The Medical Genomics Reference Bank (MGRB) cohort of healthy older Australians sequenced using ~30x Illumina HiSeq X at the Garvan Institute, Sydney. [link](https://www.biorxiv.org/content/10.1101/473348v1).                                                                                                                                                                                         |
| **MGRB_frequency**            | Represents the frequency of this variant in the MGRB cohort.                                                                                                                                                                                                                                                                                                                                            |
| **MGRB_FILTER**               | Represents the FILTER status of the variant in the MGRB cohort using GATK HaplotypeCaller.                                                                                                                                                                                                                                                                                                              |
| **MGRB_AC**                   | The allele count in the MGRB cohort; the number of individuals with this variant. Interpret this as 2x the number of individuals in the MGRB control cohort.                                                                                                                                                                                                                                            |
| **MGRB_AN**                   | The total number of alleles called in the MGRB cohort. Interpret this as 2x the number of individuals in the MGRB control cohort.                                                                                                                                                                                                                                                                       |
| **phylotree_haplotype**       | The haplotypes that the variant is known to contribute to. [link](http://www.phylotree.org/index.htm)                                                                                                                                                                                                                                                                                                   |
| **MitoTip_score**             | In silico tool for predicting the pathogenicity of novel mitochondrial tRNA variants. Score ranges from 0 to 21.8. [MitoTip Info](https://www.mitomap.org/foswiki/bin/view/MITOMAP/MitoTipInfo)                                                                                                                                                                                                         |
| **MitoTip_percentile**        | The percentile of this variant's MitoTip score compared to all possible variants in tRNA. 100% is most pathogenic, 0% is least.                                                                                                                                                                                                                                                                         |
| **MitoTip_interpretation**    | Interpretation based on MitoTip score percentile: upper 25th percentile likely pathogenic, 50-75th percentile possibly pathogenic, 25-50th percentile possibly benign, bottom 25th percentile likely benign. [MitoTip Info](https://mitomap.org/foswiki/bin/view/MITOMAP/MitoTipInfo)                                                                                                                   |
| **anticodon**                 | TRUE if the variant falls in an anticodon of a tRNA.                                                                                                                                                                                                                                                                                                                                                    |
| **disease_mitomap**           | The disease annotation for a particular variant, as annotated by MITOMAP.                                                                                                                                                                                                                                                                                                                               |
| **status_mitomap**            | Strength of evidence supporting the MITOMAP annotation: strongest evidence is Confirmed (Cfrm). [MITOMAP MutationsRNACfrm](https://mitomap.org/foswiki/bin/view/MITOMAP/MutationsRNACfrm)                                                                                                                                                                                                               |
| **GenBank_frequency_mitomap** | The frequency of GenBank records containing this variant, as a proxy for the population frequency. Confirmed pathogenic variants have frequency 0.0%.                                                                                                                                                                                                                                                   |
| **allele_frequency_mitomap**  | Derived from 32,059 GenBank full-length sequences with size greater than 15.4kbp. Range is 0-1.                                                                                                                                                                                                                                                                                                         |
| **highest_vep_impact**        | Highest VEP impact from all annotations for the variant. [VEP Info](http://www.ensembl.org/info/docs/tools/vep/index.html)                                                                                                                                                                                                                                                                              |
| **POS**                       | Variant falls in the blacklist of positions: MT:302-319, MT:3105-3108.                                                                                                                                                                                                                                                                                                                                  |
| **SBR**                       | For all alleles RO > 15 and (SBR > 0.9 or SBR < 0.1).                                                                                                                                                                                                                                                                                                                                                   |
| **SBA**                       | For all alleles AO > 15 and (SBA > 0.9 or SBA < 0.1).                                                                                                                                                                                                                                                                                                                                                   |
| **MQMR**                      | For all alleles MQMR<30                                                                                                                                                                                                                                                                                                                                                                                 |
| **AQR**                       | For all alleles AQR<20                                                                                                                                                                                                                                                                                                                                                                                  |
| **SBR_INFO**                  | For all alleles RO > 15 and (SBR > 0.9 or SBR < 0.1).                                                                                                                                                                                                                                                                                                                                                   |
| **SBA_INFO**                  | For all alleles AO > 15 and (SBA > 0.9 or SBA < 0.1).                                                                                                                                                                                                                                                                                                                                                   |
| **AQA_FORMAT**                | Average base quality of the alternate reads, AQA=QA/AO.                                                                                                                                                                                                                                                                                                                                                 |
| **AQR_FORMAT**                | Average base quality of the reference reads, AQR=QR/RO.                                                                                                                                                                                                                                                                                                                                                 |
| **VAF_FORMAT**                | Allele frequency in the range (0,1] - the ratio of the number of alternate reads to reference reads.                                                                                                                                                                                                                                                                                                    |
| **q_FORMAT**                  | Phred-scaled binomial probability of seeing AO reads from DP, assuming a noise floor of p=0.002.                                                                                                                                                                                                                                                                                                        |
